### PR TITLE
chore: Allow partial configuration file

### LINKED ISSUE #105: Allow partial configuration file where some configuration keys may be missing
**Is your feature request related to a problem? Please describe.**

If we create a partial configuration file where only some configuration options are defined, gpg-tui errors out on launch with a message similar to this:

```
Error: TOML parse error at line 1, column 1
  |
1 | [general]
  | ^^^^^^^^^
missing field `splash`
```
Gpg-tui currently doesn't accept a configuration file unless all configuration keys are defined.

**Describe the solution you'd like**

Here are some of the problems that arise from requiring the user to define all configuration keys:

- A user might wish to retain the default configuration for most of configuration keys, and override just a few of them. It's a hassle to find out the appropriate values for every key.
  - The values identified by the user for a key may be wrong - even though they may not want to override it. This happened to me. I took default values from the man pages. *However, gpg-tui was blank when it was launched. I'm still unable to figure out which key caused that*.
- If an update of gpg-tui adds a configuration key, the user will be forced to add it to the config file manually. The application won't launch without it.

The appropriate solution would be to assume default values for all keys and then allow the user to selectively override them in the config file.

**Describe alternatives you've considered**

One possibility is to define all the default values in a system-wide config file (somewhere in /etc or /usr/share). Read this file first. Then read the user config file (from ~/.config) and overwrite values for those variables that are defined in it. 

**Additional context**

None

### PR DESCRIPTION
**Pull request by PThorpe92**
## Description
This PR allows for the mandatory args that need to be generated from the config file, if the config file is found, be filled in with default arguments should they not be included in the found config file.

Please let me know if I missed anything, and thank you for the awesome program :+1: 

## Motivation and Context

closes #105 

## How Has This Been Tested?

I tested this locally on my machine by placing a config file in ~/.config/gpg-tui and 
filling adding and removing parts to make sure the program will still start.

I also wrote a test `test_args_partial_config` and included a partial conifg file `gpg-tui-part` in the config directory of the repo, and ran assertions for the missing arguments.

## Output (if appropriate):

The program will still start, despite fields missing in the configuration file.

PREV:
image
AFTER:
Well it works :+1: a screenshot of the program running is probably overkill.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other 

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -26,15 +26,15 @@ pub struct Config {
 #[derive(Debug, Default, Serialize, Deserialize)]
 pub struct GeneralConfig {
```
Comment (orhun) [inline src/config.rs:27]: Any chance that we can make the sections (`[general]` and `[gpg]`) optional as well?

File: src/config.rs [hunk 2]
```diff
@@ -163,20 +163,43 @@ impl Config {
 
 	/// Update the command-line arguments based on configuration.
 	pub fn update_args(&self, mut args: Args) -> Args {
-		args.armor = self.gpg.armor;
-		args.splash = self.general.splash;
+		let default_color = Color::from("gray");
+		let default_style = Style::from_str("plain", true).unwrap_or_default();
+		let default_file_exp = String::from("xplr");
```
Comment (orhun) [inline src/config.rs:168]: Can you use constants for these values?

**Review by orhun — CHANGES_REQUESTED** [review]
Thanks for picking this up! 💖 

I just had some comments for improvement.

File: config/gpg-tui-part.toml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+[general]
```
[the remark this answers is not in the document]
Comment (PThorpe92) [reply config/gpg-tui-part.toml:1]: Yeah for sure I do agree about it being confusing, I was just semi-unsure about how to write proper tests for something like this. Normally that file may have been included in a tests/ type directory. How would you like me to go about the test? Maybe create a temp file, write
```
[general]
splash = true
```
to it, and then run the test on that?

[the remark this answers is not in the document]
Comment (orhun) [reply config/gpg-tui-part.toml:1]: Yeah I think that approach is better. You can possibly write it to a temporary directory as well.

File: src/config.rs (hunk 2 above)
Comment (PThorpe92) [reply src/config.rs:168]: Like do you want lazy_static/and get those Color + Style variables themselves made constant? or do you want just the &str values made constant and put into the runtime functions there?

Comment (PThorpe92) [reply src/config.rs:168]: I would assume the latter actually now that I think about it.

**Comment by PThorpe92** [comment]
Take a look and let me know what you think, by making those both Option<T>, it did add some complexity when setting the default values but I added the two tests, one for each section, with the opposing one missing and then tested in both cases for each one.

I also fixed the unrelated clippy warnings that somehow still made the CI fail :eyes:

File: Cargo.lock
Comment (orhun) [inline Cargo.lock:1]: Are the changes in `Cargo.lock` necessary?

File: src/config.rs [hunk 4]
```diff
@@ -163,49 +174,131 @@ impl Config {
...
+		if let Ok(config) = Config::parse_config(&tmp_path.to_string_lossy()) {
+			let args = config.update_args(Args::default());
+			// [general]
+			assert_eq!(args.splash, true); // supplied
+			assert_eq!(args.tick_rate, 250_u64);
+			// [gpg]
+			assert_eq!(args.armor, false);
+			assert_eq!(args.default_key, None);
+		}
+		temp_file.flush()?;
```
Comment (orhun) [inline src/config.rs:275]: What's the benefit of using `flush` here? Is it really needed?

**Review by orhun — COMMENTED** [review]
Hey, overall it looks good. I just pushed e89b2be18b4ad3dd12d9ac356ef0a8a4f99f5d41 for cleanup.

Also I have 2 questions, then we can merge 🐻

File: src/config.rs (hunk 4 above)
Comment (PThorpe92) [reply src/config.rs:275]: Ah, you are correct as I forgot to remove that when I changed from write to write_all

File: Cargo.lock
Comment (PThorpe92) [reply Cargo.lock:1]: I am honestly not sure how/why this would have been generated, as I did not add/change any dependencies. Very strange.

Comment (orhun) [reply Cargo.lock:1]: Ah, it might be because I updated some dependencies in the master branch. Let's ignore it.

**Review by orhun — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**